File: ctci/python/ch1-arrays-and-strings/q6-string-compression/Question.py
# one iteration with neighbor character comparison
def compress(s):
    compressed_list = []
    count_consecutive = 1
    for i in xrange(len(s)):
        if i + 1 < len(s) and s[i] == s[i + 1]:
            count_consecutive += 1
        else:
            compressed_list.append(s[i])
            compressed_list.append(str(count_consecutive))
            count_consecutive = 1
    compressed = ''.join(compressed_list)
    return s if len(compressed) >= len(s) else compressed

File: ctci/java/ch1-arrays-and-strings/q6_string_compression/Question.java
package q6_string_compression;

/**
 * Created by jjlee on 7/12/16.
 */
public class Question {
    /* one iteration with neighbor character comparison */
    public static String compress(String s) {
        StringBuilder sb = new StringBuilder();
        int countConsecutive = 1;
        for (int i = 0; i < s.length(); i++) {
            if (i + 1 < s.length() && s.charAt(i) == s.charAt(i + 1)) {
                countConsecutive++;
            } else {
                sb.append(s.charAt(i));
                sb.append(countConsecutive);
                countConsecutive = 1;
            }
        }
        return sb.length() >= s.length() ? s : sb.toString();
    }
}
